### ISSUE TITLE
Specify concurrency semantics

### ISSUE THREAD

**Issue by erikgrinaker** [tendermint/iavl#241]
I don't believe IAVL is currently concurrency-safe. We should document this, and also specify whether e.g. an `ImmutableTree` is safe for reads while the corresponding `MutableTree` is mutated. See e.g. #240.

Related to https://github.com/tendermint/tm-db/issues/62.

**Comment by mvdan** [comment]
Here's a follow-up question. I understand that, as per https://github.com/tendermint/iavl/issues/240#issuecomment-625738226, it's non-safe to do a `tree.Set` concurrently with a `tree.ImmutableTree.Get`.

However, is it safe to do those calls concurrently if we use `GetImmutable` instead of directly accessing the same `ImmutableTree` pointer?

```
immTree := tree.GetImmutable(currentVersion)
go tree.Set(...) // concurrent write
go immTree.Get(...) // concurrent read
```

I see that `GetImmutable` doesn't return the same `*ImmutableTree`, though I understand it does share data underneath. So I'm not sure if the last bit of sample code would still be considered racy or non-safe.

**Comment by erikgrinaker** [comment]
> However, is it safe to do those calls concurrently if we use GetImmutable instead of directly accessing the same ImmutableTree pointer

I think that's true. `MutableTree.ImmutableTree` probably shouldn't have been exported, I think it's just an unfortunate side-effect of Go's type embedding. I ran some tests here with the Go race detector, and the tree returned by `GetImmutable()` does appear to be safe for concurrent use, at least in the cases I tried (simple set/get with version saving).
